### ISSUE TITLE
Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/jquery-file-upload/form.php on line 23

### ISSUE THREAD

**Issue by akhilleusuggo** [WWBN/AVideo-Encoder#200]
```
encoder::sendFile reading file from /var/www/html/YouPHPTube-Encoder/videos/5_tmpFile_converted.zip
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Warning:  array_multisort(): Array sizes are inconsistent in /var/www/html/YouPHPTube-Encoder/view/jquery-file-upload/form.php on line 21
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/jquery-file-upload/form.php on line 23
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/jquery-file-upload/form.php on line 23
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/jquery-file-upload/form.php on line 23
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/jquery-file-upload/form.php on line 23
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/jquery-file-upload/form.php on line 23
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/jquery-file-upload/form.php on line 23
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Warning:  array_multisort(): Array sizes are inconsistent in /var/www/html/YouPHPTube-Encoder/view/index.php on line 398
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 400
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 400
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 400
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 400
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 400
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 400
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Warning:  array_multisort(): Array sizes are inconsistent in /var/www/html/YouPHPTube-Encoder/view/index.php on line 449
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 451
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 451
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 451
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 451
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 451
[15-Dec-2019 10:36:05 Europe/Helsinki] PHP Notice:  Undefined property: stdClass::$hierarchyAndName in /var/www/html/YouPHPTube-Encoder/view/index.php on line 451
```

**Comment by DanielnetoDotCom** [comment]
did you make a git pull?

looks like you are missing some update

**Comment by akhilleusuggo** [comment]
I just downloaded it . I mean , I mailed you if you remember about the Nvenc , and I was trying things at home . And I'm getting this error from the home encoder , on the home server.
Btw , with nvenc , videos of 10mn got encoded in 20-30s . It's taking longer time to the packing process and sending then encoding .
I'll get pull one more time and let you know . Thank you for the reply

**Comment by DanielnetoDotCom** [comment]
hummm, good to know!! thanks

**Comment by markginter31** [comment]
akhilleusuggo -- I'm pretty novice -- and am trying to set the correct nvenc parameters in the encoder.  Would you mind sharing your ffmpeg encorder line?  Thanks.

**Comment by akhilleusuggo** [comment]
@markginter31 You haven't mentioned me , so I've got no notification . Sorry about that . Do you still need help with that ?

**Comment by markginter31** [comment]
I do still need help. Thanks for getting back to me.

**Comment by akhilleusuggo** [comment]
What GPU do you use ? Because it will vary so much . 
And What encoder ? Do you use the default encoder or the HLS ?
I'm assuming you're having the encoder at home , right ?

**Comment by markginter31** [comment]
I have an nvidia quadro p400.  I'm using arch linux since its easy to
install and update ffmpeg with nvenc support.  I'm self hosting the encoder
on a private network.

**Comment by akhilleusuggo** [comment]
@markginter31 You haven't answer all my question . It's very important to know . Because for example p400 does support only 2 of concurrent sessions

**Comment by markginter31** [comment]
Let me try again. 😁  Gpu is quadro p400. Encoder is self hosted default
(using ffmpeg).  Not hls.  Its on a private network.  I'm looking to
offload the encoding onto the gpu when I upload a video.  I'm new to all
this. Thanks for your time and patience!

**Comment by akhilleusuggo** [comment]
I was just asking about that . FFmpeg does support it btw xD 
I'm saying this , because if you're going to use it for HLS you'll need the open source nvidia drivers , since the official one has a limitation of 2 per sessions .
There are many ways , basically you'll be playing with the Ram and vRam .
For the default one you can use this line for the start .
Some explanation : 
`-hwaccel_device` :  This one is used to define witch GPU device to use , in case you have more then 1
`-threads 4` : I recomend not to use more then 4 from personal experience to reduce delay betwenn Ram and CPU
`-hwaccel cuvid & h264_cuvid` : This 2 parameters are for decoding
`h264_nvenc` : For encoding ( like in libx264 ) , but this one using nvenc
`-vf scale_npp & -vf scale` : To use this option you'll need to enable `--npp_scale` , compile it for faster scaling using nvenc . If you can't you can use software scaling instead ( gonna use your CPU ) . This one makes a huge difference on transcoding .
`-qp:v ` : This option replace ( not really ) the cbr that you can use on software encoding .
`-rc:v ` : This parameter is optional , but it may be needed . On nvenc there's no `cbr` option , and you'll always need to set a bitrate . You can remove this option if you want .
`-preset` : Possible values : `default` , `slow` , `medium` , `fast` , `hp` ( high performance ), `hq` ( high quality ) , `ll` ( low latency ) , `llhq` ( low latency high quality ) , `llhp` ( low latency high performance ) , `lossless` , `losslesshp` .

**Variant 1**
`ffmpeg -threads 4 -vsync 0 -hwaccel cuvid -hwaccel_device 0 -c:v h264_cuvid -i {$pathFileName} -c:v h264_nvenc -vf scale_npp=-2:1080 -b:v 4500k -maxrate 6000k -bufsize 8000k -profile:v main -level:v 4.2 -rc:v constqp -qp:v 32 -preset slow -rc-lookahead:v 32 -bf:v 3 -coder:v cabac -b:a 128k {$destinationFile}
`
**Variant 2**
`ffmpeg -threads 4 -hwaccel cuvid -hwaccel_device 0 -c:v h264_cuvid -i {$pathFileName} -filter:v hwupload_cuda,scale_npp=w=1920:h=1080:interp_algo=lanczos -c:v h264_nvenc -b:v 4500k -maxrate 6000k -bufsize 8000k -profile:v main -level:v 4.2 -rc:v vbr_hq -rc-lookahead:v 32 -bf:v 3 -coder:v cabac -b:a 128k {$destinationFile}`

This second one is the best ( from my experience , but you need find the way to make it work ) . Basically  the  `hwupload_cuda,scale_npp` you're uploading the video data to the GPU memory , and mixing it with CPU in case your GPU doesn't support the decoded video uploaded .

I could give you much more parameters , but I don't know if you're going to use them . Sadly your GPU doesn't support B-frames and you may find some problem if you upload videos with 10 bits . 

For fast encoding , I recommend you 3 encoding server . ( you can use all of them on the same system ) . 2 encoding server with GPU ( your GPU can handle 2 encoding at the same time ) , and one more with CPU in case some coded are not supported by your GPU .

Cheers

**Comment by markginter31** [comment]
Wow! Thanks!  Once I can try this out I'll let you know how it goes.
Thanks again!
